### ISSUE TITLE
parallel execution of tests

### ISSUE THREAD

**Issue by cogitatio** [Behat/Behat#159]
Hi!

I am quite interested to get results from tests quickly. So we are currently reaching a point, where we are splitting tests into groups to run them in parallel. But execution times are growing.

What about a central feature to let behat fork testrunners, so we can run tests in parallel? I seriously would love that.

Would need an "@exclusive" tag to run a test alone or a "@group a" which would not run grouped tests in parallel. But we are closing in on 30m tests, which is much tooo long for my taste. ;)

Best regards
Falk

**Comment by rajeshtaneja** [comment]
+1
@neochief  Any plans to update ParallelRunner to 3.x ?

**Comment by eddiejaoude** [comment]
My 2 cents...

Robo Task Runner is another option http://robo.li/tasks.html#parallelexec 

Codeception has a good usage example: http://codeception.com/docs/12-ParallelExecution

eg.

```php
    function parallelAll()
    {
        $this->parallelSplitTests();
        $result = $this->parallelRun();
        $this->parallelMergeResults();
        return $result;
    }
```

--- 

@fantasticjamieburns great command for pure maximum speed :+1:

**Comment by rajeshtaneja** [comment]
Thanks @eddiejaoude 

I am trying this now. Nice framework :+1:

**Comment by timiTao** [comment]
@eddiejaoude first link is not working...

**Comment by eddiejaoude** [comment]
@timiTao Robo have updated their website & I can't find their equivalent docs for that page. The 2nd link has an example of how to use Robo in parallel. 

It might be work logging an issue/ticket with Robo

**Comment by jwfuller** [comment]
http://robo.li/tasks/Base/#parallelexec

**Comment by timiTao** [comment]
@jwfuller  thaks a lot :) :+1:

**Comment by jhenya** [comment]
Hi there
fantasticjamieburns, 
Did you try it?
"Caveat: this approach will not aggregate the total passed/skipped/failed for reporting purposes."  commented on 1 Jul 2013

**Comment by everzet** [comment]
I'm closing this issue as this will not happen anymore. There are easier ways of achieving parallel execution. I also reached conclusion that parallelisation of tests creates a trade-off you'd never want to have in your tests - trading stability for speed. The best way to improve speed of your tests is following a testing pyramid.

**Comment by kandelyabre** [comment]
you can try extension https://packagist.org/packages/tonicforhealth/behat-parallel-scenario

**Comment by loostro** [comment]
Also see [liuggio/fastest](https://github.com/liuggio/fastest#behat2-extension)

**Comment by jhenya** [comment]
I implement parallel execution it with xargs and GNU Parallel and implement consolidated report based on BehatHtmlFormatterPlugin extension. Please see details here 
http://parallelandvisualtestingwithbehat.blogspot.com/p/blog-page.html

**Comment by seyfer** [comment]
@jhenya will not work to run tests just in parallel since with one DB parallel changes via UI lead to other tests to fail.

@ioleo thanks, so far the most advanced tool. But for me it failed to run, I have created an issue.

@neochief tried this one also. Outdated for Behat 3. Tried to fix it, but doesn't work now. 
If anyone can have a look, would be great.

Thanks all for sharing.
What worked for me so far is not paralleling, but running chrome without Selenium in headless mode. It 2-3x times faster. But without VNC.
